### ISSUE TITLE
AoT compiling fails on Windows - building conda package

### ISSUE THREAD

**Issue by msarahan** [sass/dart-sass#1780]
Greetings, and I apologize if this is a rehash of an old issue. I have looked but not found an answer.

I am trying to compile dart-sass for the conda platform at conda-forge: https://github.com/conda-forge/dart-sass-feedstock. I am mostly following the process from the chocolatey build: https://community.chocolatey.org/packages/sass#files

I am using the Dart SDK version 2.17.6, and downloading the .tar.gz file from the github tag here. In the root of that extracted tarball, I issue:

```
dart compile exe bin\sass.dart -Dversion=1.54.4 -Ddart-version=2.17.6 -o %PREFIX%\Library\bin\sass.exe 
```
which ends with:
```
Unhandled exception:
Crash when compiling null,
at character offset null:
Bad state: No element
#0      List.single (dart:core-patch/growable_array.dart:353:22)
#1      ClassBuilderImpl.buildTypeWithBuiltArguments (package:front_end/src/fasta/builder/class_builder.dart:324:44)
#2      ClassBuilderImpl.buildType (package:front_end/src/fasta/builder/class_builder.dart:335:12)
#3      NamedTypeBuilder._declarationBuildType (package:front_end/src/fasta/builder/named_type_builder.dart:390:27)
#4      NamedTypeBuilder._buildInternal (package:front_end/src/fasta/builder/named_type_builder.dart:398:28)
#5      NamedTypeBuilder.build (package:front_end/src/fasta/builder/named_type_builder.dart:382:22)
#6      FunctionTypeBuilder._buildInternal (package:front_end/src/fasta/builder/function_type_builder.dart:93:21)
#7      FunctionTypeBuilder.build (package:front_end/src/fasta/builder/function_type_builder.dart:88:22)
#8      FormalParameterBuilder.build (package:front_end/src/fasta/builder/formal_parameter_builder.dart:149:35)
#9      SourceFunctionBuilderImpl.buildFunction (package:front_end/src/fasta/source/source_function_builder.dart:346:48)
#10     SourceFactoryBuilder.build (package:front_end/src/fasta/source/source_factory_builder.dart:143:5)
#11     SourceFactoryBuilder.buildMembers (package:front_end/src/fasta/source/source_factory_builder.dart:134:21)
#12     SourceClassBuilder.build.buildBuilders (package:front_end/src/fasta/source/source_class_builder.dart:183:16)
#13     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:614:13)
#14     ConstructorScope.forEach (package:front_end/src/fasta/scope.dart:604:11)
#15     SourceClassBuilder.build (package:front_end/src/fasta/source/source_class_builder.dart:212:22)
#16     SourceLibraryBuilder.buildBuilder (package:front_end/src/fasta/source/source_library_builder.dart:3025:31)
#17     SourceLibraryBuilder.build (package:front_end/src/fasta/source/source_library_builder.dart:1025:7)
#18     SourceLoader.buildComponent (package:front_end/src/fasta/source/source_loader.dart:2016:32)
#19     KernelTarget.buildOutlines.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:475:14)
<asynchronous suspension>
#20     withCrashReporting (package:front_end/src/fasta/crash.dart:122:12)
<asynchronous suspension>
#21     _buildInternal (package:front_end/src/kernel_generator_impl.dart:134:7)
<asynchronous suspension>
#22     withCrashReporting (package:front_end/src/fasta/crash.dart:122:12)
<asynchronous suspension>
#23     kernelForProgramInternal.<anonymous closure> (package:front_end/src/api_prototype/kernel_generator.dart:62:29)
<asynchronous suspension>
#24     kernelForProgramInternal (package:front_end/src/api_prototype/kernel_generator.dart:61:10)
<asynchronous suspension>
#25     kernelForProgram (package:front_end/src/api_prototype/kernel_generator.dart:50:11)
<asynchronous suspension>
#26     compileToKernel (package:vm/kernel_front_end.dart:389:22)
<asynchronous suspension>
#27     runCompiler (package:vm/kernel_front_end.dart:292:19)
<asynchronous suspension>
#28     main (file:///C:/b/s/w/ir/cache/builder/sdk/pkg/vm/bin/gen_kernel.dart:34:19)
<asynchronous suspension>


#0      List.single (dart:core-patch/growable_array.dart:353:22)
#1      ClassBuilderImpl.buildTypeWithBuiltArguments (package:front_end/src/fasta/builder/class_builder.dart:324:44)
#2      ClassBuilderImpl.buildType (package:front_end/src/fasta/builder/class_builder.dart:335:12)
#3      NamedTypeBuilder._declarationBuildType (package:front_end/src/fasta/builder/named_type_builder.dart:390:27)
#4      NamedTypeBuilder._buildInternal (package:front_end/src/fasta/builder/named_type_builder.dart:398:28)
#5      NamedTypeBuilder.build (package:front_end/src/fasta/builder/named_type_builder.dart:382:22)
#6      FunctionTypeBuilder._buildInternal (package:front_end/src/fasta/builder/function_type_builder.dart:93:21)
#7      FunctionTypeBuilder.build (package:front_end/src/fasta/builder/function_type_builder.dart:88:22)
#8      FormalParameterBuilder.build (package:front_end/src/fasta/builder/formal_parameter_builder.dart:149:35)
#9      SourceFunctionBuilderImpl.buildFunction (package:front_end/src/fasta/source/source_function_builder.dart:346:48)
#10     SourceFactoryBuilder.build (package:front_end/src/fasta/source/source_factory_builder.dart:143:5)
#11     SourceFactoryBuilder.buildMembers (package:front_end/src/fasta/source/source_factory_builder.dart:134:21)
#12     SourceClassBuilder.build.buildBuilders (package:front_end/src/fasta/source/source_class_builder.dart:183:16)
#13     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:614:13)
#14     ConstructorScope.forEach (package:front_end/src/fasta/scope.dart:604:11)
#15     SourceClassBuilder.build (package:front_end/src/fasta/source/source_class_builder.dart:212:22)
#16     SourceLibraryBuilder.buildBuilder (package:front_end/src/fasta/source/source_library_builder.dart:3025:31)
#17     SourceLibraryBuilder.build (package:front_end/src/fasta/source/source_library_builder.dart:1025:7)
#18     SourceLoader.buildComponent (package:front_end/src/fasta/source/source_loader.dart:2016:32)
#19     KernelTarget.buildOutlines.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:475:14)
<asynchronous suspension>
#20     withCrashReporting (package:front_end/src/fasta/crash.dart:122:12)
<asynchronous suspension>
#21     _buildInternal (package:front_end/src/kernel_generator_impl.dart:134:7)
<asynchronous suspension>
#22     withCrashReporting (package:front_end/src/fasta/crash.dart:122:12)
<asynchronous suspension>
#23     kernelForProgramInternal.<anonymous closure> (package:front_end/src/api_prototype/kernel_generator.dart:62:29)
<asynchronous suspension>
#24     kernelForProgramInternal (package:front_end/src/api_prototype/kernel_generator.dart:61:10)
<asynchronous suspension>
#25     kernelForProgram (package:front_end/src/api_prototype/kernel_generator.dart:50:11)
<asynchronous suspension>
#26     compileToKernel (package:vm/kernel_front_end.dart:389:22)
<asynchronous suspension>
#27     runCompiler (package:vm/kernel_front_end.dart:292:19)
<asynchronous suspension>
#28     main (file:///C:/b/s/w/ir/cache/builder/sdk/pkg/vm/bin/gen_kernel.dart:34:19)
<asynchronous suspension>
Error: AOT compilation failed
Generating AOT kernel dill failed!
```

As a side note, if anyone here would like to be a "maintainer" on the feedstock repo, let me know. Maintaining the feedstock doesn't imply responsibility so much as it means that you have the power to merge PRs there (for example, to merge PRs from the bot that update to the latest version).

**Comment by nex3** [comment]
This is definitely a bug in Dart. I recommend filing a bug on dart-lang/sdk. I'm not sure specifically what's happening there; it doesn't seem to reproduce on my Linux machine.

If you're compiling Dart Sass from source, you should generally use the Grinder tasks rather than invoking `dart` yourself. You can see a list of them by running `dart run grinder --help` in the repo root. In this case, you probably want `pkg-standalone-windows-x64`.

**Comment by msarahan** [comment]
Sorry to keep bothering you. The grinder command doesn't work for me. Again, I'm starting from the source tarball from https://github.com/sass/dart-sass/archive/refs/tags/1.54.4.tar.gz, and with the dart SDK from https://storage.googleapis.com/dart-archive/channels/stable/release/2.17.6/sdk/dartsdk-windows-x64-release.zip

```
(%BUILD_PREFIX%) %SRC_DIR%>dart pub get
Resolving dependencies...
+ _fe_analyzer_shared 38.0.0 (46.0.0 available)
+ analyzer 3.4.1 (4.6.0 available)
+ archive 3.3.1
+ args 2.3.1
+ async 2.9.0
+ boolean_selector 2.1.0
+ charcode 1.3.1
+ checked_yaml 2.0.1
+ cli_pkg 2.1.4
+ cli_repl 0.2.2
+ cli_util 0.3.5
+ collection 1.16.0
+ convert 3.0.2
+ coverage 1.5.0
+ crypto 3.0.2
+ csslib 0.17.2
+ dart_style 2.2.3
+ dartdoc 5.1.2 (6.0.1 available)
+ file 6.1.3
+ freezed_annotation 1.1.0 (2.1.0 available)
+ frontend_server_client 2.1.3
+ glob 2.1.0
+ grinder 0.9.2
+ html 0.15.0
+ http 0.13.5
+ http_multi_server 3.2.1
+ http_parser 4.0.1
+ io 1.0.3
+ js 0.6.4
+ json_annotation 4.6.0
+ lints 1.0.1 (2.0.0 available)
+ logging 1.0.2
+ markdown 5.0.0 (6.0.0 available)
+ matcher 0.12.12
+ meta 1.8.0
+ mime 1.0.2
+ node_interop 2.1.0
+ node_preamble 2.0.1
+ oauth2 2.0.0
+ package_config 2.1.0
+ path 1.8.2
+ petitparser 5.0.0
+ pool 1.5.1
+ pub_api_client 2.2.1
+ pub_semver 2.1.1
+ pubspec 2.3.0
+ pubspec_parse 1.2.0
+ quiver 3.1.0
+ shelf 1.3.2
+ shelf_packages_handler 3.0.1
+ shelf_static 1.1.1
+ shelf_web_socket 1.0.2
+ source_map_stack_trace 2.1.0
+ source_maps 0.10.10
+ source_span 1.9.1
+ stack_trace 1.10.0
+ stream_channel 2.1.0
+ stream_transform 2.0.0
+ string_scanner 1.1.1
+ term_glyph 1.2.1
+ test 1.21.4
+ test_api 0.4.12
+ test_core 0.4.16
+ test_descriptor 2.0.0
+ test_process 2.0.2
+ tuple 2.0.0
+ typed_data 1.3.1
+ uri 1.0.0
+ vm_service 9.3.0
+ watcher 1.0.1
+ web_socket_channel 2.2.0
+ webkit_inspection_protocol 1.1.0
+ xml 5.4.1 (6.1.0 available)
+ yaml 3.1.1
Changed 74 dependencies!

(%BUILD_PREFIX%) %SRC_DIR%>dart run grinder pkg-standalone-windows-x64 
Could not find a command named "C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\bin\snapshots\frontend_server.dart.snapshot".

Usage: dart <command|dart-file> [arguments]

Global options:
-h, --help                 Print this usage information.
-v, --verbose              Show additional command output.
    --version              Print the Dart SDK version.
    --enable-analytics     Enable analytics.
    --disable-analytics    Disable analytics.

Available commands:
  analyze    Analyze Dart code in a directory.
  compile    Compile Dart to various formats.
  create     Create a new Dart project.
  devtools   Open DevTools (optionally connecting to an existing application).
  doc        Generate API documentation for Dart projects.
  fix        Apply automated fixes to Dart source code.
  format     Idiomatically format Dart source code.
  migrate    Perform null safety migration on a project.
  pub        Work with packages.
  run        Run a Dart program.
  test       Run tests for a project.

Run "dart help <command>" for more information about a command.
See https://dart.dev/tools/dart-tool for detailed documentation.
Cannot rename file to 'C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\work\.dart_tool/pub\incremental\grinder\grinder.dart.incremental.dill', path = 'C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\work\.dart_tool/pub\incremental\grinder\tmpdc914f41\grinder.dart.incremental.dill.incremental.dill' (OS Error: The system cannot find the file specified.
, errno = 2)
FileSystemException: Cannot rename file to 'C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\work\.dart_tool/pub\incremental\grinder\grinder.dart.incremental.dill', path = 'C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\work\.dart_tool/pub\incremental\grinder\tmpdc914f41\grinder.dart.incremental.dill.incremental.dill' (OS Error: The system cannot find the file specified.
, errno = 2)
```

I noticed that this has a slash direction mixup:

`C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\work\.dart_tool/pub\incremental\grinder\tmpdc914f41\grinder.dart.incremental.dill.incremental.dill`
or more specifically:
`work\.dart_tool/pub\incremental`

I'm not having much luck figuring this out, because I can't find any source code that has these paths. For what it's worth, this path (same as before, but with reversed slash fixed) does exist, and is empty: `C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\work\.dart_tool\pub\incremental\grinder\`

The parent shell in my case is cmd.exe. I'm not sure if calling dart from powershell would behave differently, but I don't think so.

**Comment by nex3** [comment]
That also looks like either a bug in Dart or a broken installation. I recommend trying to run Dart from the official SDK package and see if that works better.

**Comment by msarahan** [comment]
I thought I was running the official SDK package: https://dart.dev/get-dart/archive

I did bundle it into a conda package, so I'll circle back around and ensure that that process hasn't messed things up.

One thing I noticed is that the path it's trying to run seems to be a result of a relative path gone one level too high:

```
(%BUILD_PREFIX%) %SRC_DIR%>dart run grinder pkg-standalone-windows-x64 
Could not find a command named "C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\bin\snapshots\frontend_server.dart.snapshot".
```

However, the root of the source code where the working directory is is at

`C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\work`
and all I have in the bin folder therein is:

```
C:\Users\msarahan\mambaforge\conda-bld\dart-sass_1660769458193\work\bin\sass.snapshot
```

In other words, there is no bin/snapshots folder in my source tree. Does that come from dart-sass or from dart?

When you say "broken installation" - are there things I could do to diagnose or verify correct behavior?
